### ISSUE TITLE
[RFC] DMD2 FSDP engine and offline trainer integration

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#535]
## Feature request

Build the **DMD2 FSDP engine and offline trainer integration** for the Qwen-Image MVP in #519, using the loss/config work in #520 and the concrete adapter/acceptance work in #543.

Following [cr-gao's review](https://github.com/verl-project/verl-omni/pull/545#issuecomment-5609385564), replace the general role-runtime/controller/executor design with a concrete DMD2 integration. Existing #546/prototypes are migration inputs, not validation of this revision.

## Motivation

DMD2 needs differentiable sampling, detached fake-stage samples and independent student/fake optimization, not another Ray lifecycle or arbitrary role/transport framework. Adapt the existing worker/engine's single-active-optimizer assumptions rather than treating DMD2 as PPO replay or an OPD request.

Only **DMD2 distribution-only** is in scope. Original DMD, GAN, causal Wan and Self-Forcing are not prerequisites.

## Proposed structure

```text
main_diffusion / existing task and resource-pool setup
  -> DistributionMatchingRayTrainer(BaseRayDiffusionTrainer)
       -> existing worker API with minimal DMD-specific integration
            -> DMDDiffusersFSDPEngine(DiffusersFSDPEngine)
                 -> Qwen architecture adapter / frozen conditioning
                 -> registered DMDLoss and fake-score denoising math
                 -> existing optimizer, offload, LoRA and checkpoint helpers
```

| Component | Responsibility |
| --- | --- |
| Trainer in `trainer/diffusion/ray_diffusion_trainer.py` | Offline lifecycle, data iteration, short `1:K` loop, counters, logging/profiling and save/resume scheduling |
| Existing `TrainingWorker`, with a narrow construction seam or thin specialization if necessary | Distributed setup, dispatch, typed config, batching/result transport and engine delegation |
| Proposed `workers/engine/fsdp/dmd_impl.py` | DMD forwards/backward, active-model/optimizer ownership, numerical update outcome, EMA and DMD-specific checkpoint state |
| Qwen pipeline adapter | Prompt encoding, latent geometry/packing, scheduler/model-input conversion and raw prediction adaptation |
| `DMDLoss` / pure math helpers | Student surrogate and detached fake-score target/loss, not execution or resource management |

Reuse `EngineRegistry` and the model registry, without requiring `DistillationPlan`, a generic controller/executor, placement graph or architecture-computer protocol.

The base `fit()` is abstract: reuse concrete offline DPO/NFT setup/services, not paired-data/reward computation or OPD's trajectory replay. Extract actual duplication rather than copying a loop and bypassing most of it.

### Worker and engine integration

- Reuse process-group, NUMA/device, profiler, ND dispatch and result-collection setup.
- Add only a narrow construction hook/thin subclass for DMD settings; no hidden optimizer fields or unrestricted runtime factory.
- Select the role/optimizer **before** train/offload contexts; restore adapter/mode state safely through exceptions and checkpoint recomputation.
- Implement the base engine's real batch/forward/loss interfaces, not PPO's trajectory loop or unsupported stubs.
- Reuse microbatch preparation/accumulation and typed configs; allow required per-role sizes without a new splitter or PPO-named DMD fields.
- Register DMD2 as a real loss mode instead of permanently bypassing invalid PPO configuration.

### Model and optimizer ownership

Recommended first path:

```text
one immutable Qwen base
  student adapter       -> student optimizer / scheduler
  fake_score adapter    -> fake optimizer / scheduler
  adapters disabled     -> frozen teacher, no optimizer
  student_ema adapter   -> smoothing state, no optimizer
```

Use `LoRAAdapterMixin` for initialization, selection, copy and EMA. Parameter sets owned by student and fake optimizers must be nonempty and disjoint; teacher/EMA parameters are excluded. Clear only the relevant gradients, assert isolation, and do not let no-grad teacher/fake scoring destroy a pending student graph.

For FSDP1 shared adapters retain `use_orig_params=true` as the tested constraint. Validate FSDP1 and FSDP2 separately with activation checkpointing; unwrapped PyTorch success is not sufficient.

Sharing is not required mathematically. Explicit model access should permit an independent-module oracle and later causal/bidirectional separation, without advertising arbitrary placement, full-fine-tuning or standalone scoring.

## Short loop and update outcomes

A completed outer cycle contains one student attempt followed by `K` fake-score attempts. Each stage obtains a fresh prompt batch in the MVP baseline; fake generation uses the current student under no-grad rather than carrying a student graph across RPCs. Data exhaustion may restart the stateful iterator; fake-stage fetches must not accidentally shorten the configured cycle budget. Record student/fake sample consumption separately.

Constrain `train_mini_batch` to one optimizer minibatch and one epoch per role call, or supply an equally explicit single-update entrypoint. Internal microbatch accumulation remains supported. The first implementation returns **0 or 1 actual successful optimizer steps**, never assumes that one `update_actor()` call is one step.

A small result record, carried by the existing TensorDict/non-tensor output mechanism, is sufficient:

```text
status:          updated | skipped
optimizer_steps: 1       | 0
skip_reason:     absent  | nonfinite_grad
metrics:         role loss / grad norm / timings / diagnostics
```

This is not a new general executor protocol. Missing fields, an unexplained zero count, a count greater than one, unknown status/reason or inconsistent rank outcomes are errors.

### Explicit numerical-skip policy

This is a proposed MVP behavior, replacing the old fatal-zero-progress controller rule:

1. Agree on non-finite-gradient skip decisions across ranks sharing the optimizer **before** an update can be applied inconsistently.
2. A recognized numerical skip clears that role's gradients and returns zero successful steps. Its scheduler and student EMA do not advance.
3. Continue the remaining scheduled attempts. Do not automatically retry a skipped stage or silently compensate with extra fake updates. The configured `1:K` is the attempt schedule; report actual successful counts and skip rates separately.
4. A successful student step triggers an eligible EMA update exactly once. Eligibility/start settings use successful student-update counts, not fake updates, microbatches or skipped attempts.
5. Any unexpected forward, data, dispatch, checkpoint or communication exception remains fatal. Do not broadly catch exceptions and label them numerical skips.

**Accounting:** `global_step` advances once after a completed outer cycle, including a cycle whose valid outcomes are all numerical skips. `student_optimizer_steps` and `fake_score_optimizer_steps` advance only by returned successful counts. Their normal no-skip relationship is `student=global_step` and `fake=K*global_step`; skipped runs need not satisfy it.

`trainer.total_training_steps` is a finite cycle budget. Checkpoint/progress/logging use cycle progress; optimizer LR schedules use their own successful-step clocks, with nominal horizons derived from the configured opportunities. This avoids both infinite zero-success loops and overwriting multiple checkpoints with an unchanged student-success index. An all-skipped run must be reported as numerically unsuccessful, not as a successful training demonstration.

This changes the previous RFC's meaning of `global_step`; identify it in checkpoint metadata and migration documentation rather than silently reinterpreting old state.

### Gradient and loss aggregation

Teacher scoring and fake-stage student generation are no-grad; only the required student graph is retained. Preserve #520's loss boundary.

Unequal final microbatches require correct sample/element weighting. Preserve matching numerator/denominator and FSDP/DP normalization before clipping, without a second loss-wrapper division. Compare physical and accumulated losses/gradients/updates with replayed randomness. Existing end-of-minibatch logic must not double-step a scheduler or advance it on a skip.

### Distributed control flow and metrics

Synchronize exits/branches that change collective counts; tensor noise may differ across samples. Preserve resolved-future/error-propagation fixes. Blocking dispatch is not automatic collective cancellation; unexpected rank failures remain fatal.

Reuse profiling for attempts/successes/skips, component timings and memory. Retain every fake stage. Durations/counts are not sample-weighted losses; nested host timers are not additive isolated GPU costs.

## Checkpoint, resume and export

Reuse existing FSDP checkpoint managers and trainer directory/publication conventions. Save each physical model once, with:

- student and fake optimizer/scheduler states;
- EMA and selected-adapter metadata;
- driver/data position and worker-local sampling RNG states;
- cycle, successful-update and skip counters;
- resolved model/config/sampling fingerprints and format/accounting version.

Use existing extra-state support or a small additional DMD payload, not an actor-only save that omits fake state or a universal checkpoint framework. Publish complete boundaries atomically; a failed save never advances the latest pointer.

Partial-cycle failures may have changed parameters: stop and restore the last complete checkpoint, without in-process retry or claimed parameter rollback. A failed restore is fatal. Old prototype formats need tested migration or clear rejection before mutation; existing artifacts remain untouched.

Default export is **student**, using its own PEFT config, lazy-materialization context and existing merged-LoRA restrictions. EMA remains training state and an optional export; teacher/fake state is never exported. Require reload validation, not a new inference service, validation-replica lifecycle or standalone transport.

## Configuration and acceptance

Use `algorithm.trainer_type=distribution_matching`, `algorithm.sample_source=offline`, model/loss algorithm `dmd2`, and the top-level `dmd:` group from #520. Leave OPD's `distillation:` group and `inline`/`one_step_off` semantics unchanged. Reject OPD enable/loss flags and unsupported methods, objectives or layouts before allocation.

- [ ] Existing worker/base-engine infrastructure is reused; no parallel generic controller, role graph or executor hierarchy is required.
- [ ] Every role call produces an explicit valid outcome, at most one optimizer update, and correct scheduler/EMA behavior.
- [ ] CPU tests cover construction, config routing, aggregation, update counts, skips, malformed results, data iteration, complete/failed saves and compatibility rejection.
- [ ] Multi-rank FSDP1/FSDP2 tests cover pending student graphs, adapter switching, inactive gradients, unequal microbatches, numerical-skip agreement and repeated cycles.
- [ ] Injected non-finite gradients do not cause fatal-controller behavior, false EMA updates, false success counts or an unbounded loop; unexpected exceptions still stop execution.
- [ ] Resume matches uninterrupted execution under controlled weights/data/RNG, including both optimizers and counters; export reload reproduces the selected student parameters.
- [ ] #543 completes real Qwen multi-GPU training and artifact validation before the MVP is advertised as runnable.
- [ ] Existing OPD/DPO/NFT/FlowGRPO regressions and config/sanity checks remain green on recorded revisions.

## Your contribution

Reuse the tested prototype code and preserve out-of-scope branches/artifacts. Deliver #543 with this runtime where useful so review includes a real consumer. Historical CPU/GPU results are not refactor acceptance; code PRs require current tests, contribution disclosures and fresh human review.

**Comment by OnePunchMonk** [comment]
Hey @NancyFyong, I saw the “help wanted” label here. Let me know if there’s anything I can help with. Happy to assist. thank you!

**Comment by NancyFyong** [comment]
> Hey [@NancyFyong](https://github.com/NancyFyong), I saw the “help wanted” label here. Let me know if there’s anything I can help with. Happy to assist. thank you!

Hi @OnePunchMonk , Thank you very much for your interest. We are currently building the framework structure; once the framework is in place, we can proceed with distillation algorithms such as DMD, DMD2, and Self-Forcing. We can collaborate further once this is ready.
